File: examples/SmartHomeDemo/src/User/SmartHome/NestThermostat.cs
using System;
using System.Collections.Generic;

namespace SmartHome;

public partial class NestThermostat
{
    private double _currentTemperature = 19.3;

    public NestThermostat(string deviceId, string displayName)
    {
        DeviceId = deviceId;
        DisplayName = displayName;
    }

    public HashSet<DeviceFeature> GetSupportedFeatures()
    {
        return
        [
            DeviceFeature.Climate,
            DeviceFeature.Automation,
        ];
    }

    public string DescribeState()
    {
        return $"target={TargetTemperature:F1}C current={ReadTemperature():F1}C";
    }

    public double ReadTemperature()
    {
        _currentTemperature = Math.Round((_currentTemperature + TargetTemperature) / 2.0, 1);
        return _currentTemperature;
    }
}

File: examples/SmartHomeDemo/cpp/user/src/SmartHome/NestThermostat.cpp
#include "SmartHome/NestThermostat.hpp"

#include <cmath>
#include <sstream>
#include <utility>

namespace SmartHome
{

NestThermostat::NestThermostat(::Crimson::Cpp::String deviceId, ::Crimson::Cpp::String displayName)
{
    SetDeviceId(std::move(deviceId));
    SetDisplayName(std::move(displayName));
}

::Crimson::Cpp::Set<DeviceFeature> NestThermostat::GetSupportedFeatures()
{
    return
    {
        DeviceFeature::Climate,
        DeviceFeature::Automation,
    };
}

::Crimson::Cpp::String NestThermostat::DescribeState()
{
    std::ostringstream builder;
    builder.setf(std::ios::fixed);
    builder.precision(1);
    builder << "target=" << GetTargetTemperature() << "C current=" << ReadTemperature() << "C";
    return builder.str();
}

double NestThermostat::ReadTemperature()
{
    currentTemperature_ = std::round((currentTemperature_ + GetTargetTemperature()) / 2.0 * 10.0) / 10.0;
    return currentTemperature_;
}

}
